File: 2015/02_I_Was_Told_There_Would_Be_No_Math/Day02.kt
// Day 2: I Was Told There Would Be No Math
// https://adventofcode.com/2015/day/2

import java.nio.file.Files
import java.nio.file.Paths

fun readInput(name: String): String {
    return Files.readString(Paths.get("$name.txt")).trim()
}

data class Box(val l: Int, val w: Int, val h: Int) {
    fun calculateWrappingPaper(): Int {
        val sides = listOf(l * w, w * h, h * l)
        val smallestSide = sides.minOrNull() ?: 0
        return 2 * sides.sum() + smallestSide
    }

    fun calculateWrappingRibbon(): Int {
        val maxSide = listOf(l, w, h).maxOrNull() ?: 0
        val otherSides = l + w + h - maxSide
        val perimeter = 2 * otherSides
        val volume = l * w * h
        return perimeter + volume
    }
}

fun prepareData(input: String): List<Box> {
    return input.split("\r\n").map { box ->
        val (l, w, h) = box.split("x").map { it.toInt() }
        Box(l, w, h)
    }
}

fun part1(input: String): Int {
    val boxes = prepareData(input)
    return boxes.sumOf { it.calculateWrappingPaper() }
}

fun part2(input: String): Int {
    val boxes = prepareData(input)
    return boxes.sumOf { it.calculateWrappingRibbon() }
}

fun main() {
    val day = "Day02"
    val input = readInput(day)
    val inputTest = readInput("${day}_test")

    println("Part 1: ${part1(input)}")
    println("Part 1: ${part1(inputTest)}")

    println("Part 2: ${part2(input)}")
    println("Part 2: ${part2(inputTest)}")
}

File: 2015/02_I_Was_Told_There_Would_Be_No_Math/Day02.js
// Day 2: I Was Told There Would Be No Math
// https://adventofcode.com/2015/day/2

const utils = require('../../utils');

function prepareData(input) {
  return input.split('\r\n').map((box) => box.split('x').map(Number));
}

function calculateWrappingPaper(box) {
  const [l, w, h] = box;
  const sides = [l * w, w * h, h * l];
  const smallestSide = Math.min(...sides);
  return 2 * sides.reduce((acc, side) => acc + side, 0) + smallestSide;
}

function calculateWrappingRibbon(box) {
  const [l, w, h] = box;
  const maxSide = Math.max(l, w, h);
  const otherSides = l + w + h - maxSide;
  const perimeter = 2 * otherSides;
  const volume = l * w * h;
  return perimeter + volume;
}

function part1(input) {
  const boxes = prepareData(input);
  return boxes.reduce((total, box) => total + calculateWrappingPaper(box), 0);
}

function part2(input) {
  const boxes = prepareData(input);
  return boxes.reduce((total, box) => total + calculateWrappingRibbon(box), 0);
}

const day = 2;
const input = utils.readInputByDay(day);
const inputTest = utils.readInputByDay(day, '_test');

console.log('Part 1:', part1(input)); // 101
console.time('Part 1');
console.log('Part 1:', part1(inputTest)); // 1588178
console.timeEnd('Part 1');

console.log('Part 2:', part2(input)); // 48
console.time('Part 2');
console.log('Part 2:', part2(inputTest)); // 3783758
console.timeEnd('Part 2');
